### ISSUE TITLE
Cast is not applied when it should be

### ISSUE THREAD

**Issue by hylkevds** [jOOQ/jOOQ#8191]
### Expected behavior and actual behavior:

The documentation of Field.div states:

> If this is a numeric field, then the result is a number of the same type as this field.

Meaning that `Field<Integer>.div(Field<Double>)` results in an integer. However, PostgreSQL disagrees:
```
postgres=# select 5::integer / 2::double precision;
      ?column?      
--------------------
 2.5000000000000000
(1 row)
```

Casting the result to Integer, in order to pass it to a function that requires an Integer then fails, since jooq will eat the cast, as it thinks the result is already Integer.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

```
SelectSelectStep<Record1<Integer>> select = dslContext.select(DSL.val(7).div(2.0).cast(SQLDataType.INTEGER).mod(2));
        Record1<Integer> fetchAny = select.fetchAny();
```
Should work, but doesn't because:
`org.jooq.exception.DataAccessException: SQL [select mod((? / ?), ?)]; ERROR: function mod(double precision, integer) does not exist
`
Note that the cast is gone from the SQL.

### Versions:

- jOOQ: 3.11.7
- Java: any
- Database (include vendor): PostgreSQL 9.5
- OS: any
- JDBC Driver (include name if inofficial driver): postgresql-9.4.1212

**Comment by lukaseder** [comment]
Thank you very much for your report. There are two things that are worth noting here:

**1. Unfortunately, this behaviour isn't too standard across RDBMS.**

Take Oracle, for instance, where most people use the decimal `NUMBER` data type (and only rarely `BINARY_DOUBLE` etc.), in case of which this is only a matter of precision and scale, not of data types. It will be difficult to encode the native behaviour of all databases in a consistent way across the entire jOOQ API - notice this isn't just about divisions. All other arithmetic operations as well as many function calls are affected as well.

**2. What you're requesting is impossible to model through Java's generics**

Let's assume our universe consists only of two data types: `int` and `double`. We now have four different possible combinations for operations like the division:

- Horizontally: Dividend type
- Vertically: Divisor type
- Cell contents: Quotient type

<table>
<tr><th></th><th>int</th><th>double precision</th></tr>
<tr><th>int</th><td>int</td><td>double precision</td></tr>
<tr><th>double precision</th><td>double precision</td><td>double precision</td></tr>
</table>

This makes sense in the same way that e.g. the Java language implements automatic type promotion for what would otherwise be incompatible data types. The only reasonable thing you would have wanted to get out of a `int / double precision` expression is a `double precision` expression, and the way you'll expect this to work is by implicitly casting the `int` expression to `double precision`.

While such implicit type conversion would totally be possible in the jOOQ runtime query model, how to model that through generics? We'd need a compiler plugin that would analyse all the possible combinations of `T1` and `T2` types in a `Field<T1>.div(Field<T2>)` expression and produce the right `T3` type as a result. The Java language doesn't support that.

A workaround would be provide alternative `Field<T>` types for certain data types, such as `IntegerField extends Field<Integer>`, and then implement different method signatures for operators like `div()`. Unfortunately, this wouldn't be possible through subtyping. Also, overloading the different methods wouldn't be possible as well, because of generic type erasure. It's not possible to have:

```java
class IntegerField implements Field<Integer> {
  ...
  IntegerField div(Field<Integer>)
  DoubleField div(Field<Double>)
}
```

Even if it would be possible, [correlating subtype and parametric polymorphism will bite you in many different and difficult to foresee ways](https://blog.jooq.org/2013/06/28/the-dangers-of-correlating-subtype-polymorphism-with-generic-polymorphism/). One problem is that you will not always be able to get an `IntegerField` or a `DoubleField` type out of an arbitrary jOOQ expression, but receive the super type `Field<Integer>` or `Field<Double>` instead, where you would again have the more generic, type agnostic `div()` method.

**TL;DR:** There's no way to implement what you want in an internal DSL like jOOQ. So, the most sensible thing to do is to either require both operands to be of the same type, or to allow for the right operand to be of an arbitrary type and casting it to the type of the left operand. jOOQ chose the latter, which is more convenient. Your request shows that maybe, we should have chosen the former, to make sure people do not run into this problem where the generated SQL behaves "unexpectedly"

**Comment by lukaseder** [comment]
Oh my. I wrote all that text only to realise your report was about something entirely else 🤦‍♂️. Let's try again (I'll leave the text for future visitors)

**Comment by lukaseder** [comment]
I agree that the cast should be applied in this situation. The workaround is obviously to make sure the division operation is not of type `Integer`, manually. E.g. by casting the divident to `DOUBLE`.

Ignoring certain casts is a very old feature. I vaguely remember that jOOQ used to generate too many casts internally, in situations where they would not have been needed. I think that "optimisation" is wrong as it has side effects such as this one. When users make explicit casts (as opposed to jOOQ-internal casts), then those casts *must* appear in generated SQL.

**Comment by hylkevds** [comment]
Thanks for the two replies. I wasn't sure which of the two "issues" would be the "bug", since they are related.
Maybe would be good to add a warning to the javadoc of the functions where JOOQ can return a different Field type than the DBMS?

Since this issue is closed, is the "manual casts must not be ignored" still on the todo list, or is it already fixed in a future version?

**Comment by hylkevds** [comment]
Thinking a bit more on this, there are two possible cases with "The DB returns A, JOOQ thinks it's B"

1. You want B. The above described scenario, where JOOQ eats the cast.
2. You want A. In this case we can cast, and JOOQ would add the cast to the SQL, but it's a useless cast. In the best case it just muddies the query and is removed by the query optimiser of the DBMS.

To resolve 2, would it be an option to add a method just to tell JOOQ it has the wrong type, and correct it? Something like `Field<T> DSL.fixType(Field<?>, Class<T> correctType);`
Or does such a method already exist?

**Comment by lukaseder** [comment]
> Or does such a method already exist?

That method is called `DSL.coerce()`

**Comment by lukaseder** [comment]
Will look into fixing this through #8234

**Comment by lukaseder** [comment]
#8234 is implemented in jOOQ 3.12. It will not be backported to 3.11, as it constitutes an incompatible change of (undocumented) behaviour
